src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -77,40 +92,145 @@
+
+(defn- normalize-request
+  "Converts request to the \"one true format\" if possible.  Ensures
+  that the request :params include \"command\", \"version\", and maybe
+  \"certname\" entries, and that the entries are in the correct
+  format, i.e. spaces instead of underscores in the command name,
+  integer instead of string for the version, etc.  Ensures that
+  the :body only contains the \"payload\", as either a string or
+  stream.  The :body will be a stream unless reading the body stream
+  is unavoidable (i.e. old-style, non-param POST)."
+  [handle]
+  (fn [{:keys [body params] :as req}]
+    (let [req (if-let [cmd (params "command")]
+                (assoc-in req [:params "command"] (str/replace cmd "_" " "))
+                req)
+          req (if-let [ver (params "version")]
+                (assoc-in req [:params "version"] (Integer/parseInt ver))
+                req)
+          {:strs [command version certname]} params
+          missing-items (fn [kind command version certname]
+                          (http/error-response
+                           (str "Missing POST " kind ":"
+                                (when-not command " command")
+                                (when-not version " version")
+                                (when-not certname " certname"))))]
+      (cond
+        (and command version certname) (handle req)
+
+        (or command version certname)
+        (missing-items "parameters" command version certname)
+        
+        :else        
+        (do
+          (log/warn
+           (str "Unable to stream command posted without parameters"
+                " (loading into RAM)"))
+          (if-let [body-str (request/body-string req)]
+            (let [body (json/parse-string body-str)                    
+                  required-fields #{"command" "version" "payload"}]
+              (if-not (= required-fields (set (keys body)))
+                (http/error-response
+                 (str "Unexpected POST command fields: "
+                      (str/join " " (set/difference (set (keys body))
+                                                    required-fields))))
+                (let [certname (get-in body ["payload" "certname"])
+                      params (merge params
+                                    (select-keys body ["command" "version"])
+                                    (when certname {"certname" certname}))
+                      {:strs [command version certname]} params]
+                  ;; Enforce certname too? (currently breaks tests)
+                  (if-not (and command version)
+                    (missing-items "elements" command version true)
+                    (handle
+                     (assoc req
+                            :body (json/generate-string (body "payload"))
+                            :params (assoc params "version"
+                                           (Integer. version))))))))
```

Comment:
I *think* I just copied what was in the final version of @wkalt's patch (that this is on top of), as I thought there'd been discussion there.